go/vt/vtgate/buffer/shard_buffer.go

```diff
@@ -569,10 +569,32 @@ func (sb *shardBuffer) drain(q []*entry, err error) {
 	sb.timeoutThread.stop()
 
 	start := sb.timeNow()
-	// TODO(mberlin): Parallelize the drain by pumping the data through a channel.
+
+	// Parallelize the drain by pumping the data through a channel.
+	entryChan := make(chan *entry, len(q))
+
+	parallelism := min(bufferDrainConcurrency, len(q))
+
+	var wg sync.WaitGroup
+	wg.Add(parallelism)
+
+	for i := 0; i < parallelism; i++ {
+		go func() {
+			for _, e := range q {
+				sb.unblockAndWait(e, err, true /* releaseSlot */, true /* blockingWait */)
+			}
+
+			wg.Done()
+		}()
+	}
+
 	for _, e := range q {
-		sb.unblockAndWait(e, err, true /* releaseSlot */, true /* blockingWait */)
+		entryChan <- e
 	}
+
+	close(entryChan)
+	wg.Wait()
```

Comment:
Would it make sense to only pump the entries through the channel if `bufferDrainConcurrency` is set to a value higher than `1`? Or is the overhead negligible?

Reply:
No, I think this is an important optimization. We need a legacy codepath for when `bufferDrainConcurrency <= 1`.